### PR TITLE
Change EditPersonDescriptor parameters to be non-Optional #480

### LINKED ISSUE #480: Change EditPersonDescriptor parameters to be non-Optional
Checkstyle reports that `Optional` should not be used as a type for fields and parameters.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #480

### CODE REVIEW

**Comment by Zhiyuan-Amos** [comment]
@damithc prof did a quick fix for #480. So far it doesn't seem that the usage of `Optional` as parameters is justified, as the codebase is being littered with `Optional.of(...)` (see the code deleted).

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -91,11 +90,16 @@ private static Person createEditedPerson(ReadOnlyPerson personToEdit,
                                              EditPersonDescriptor editPersonDescriptor) {
         assert personToEdit != null;
 
-        Name updatedName = editPersonDescriptor.getName().orElseGet(personToEdit::getName);
-        Phone updatedPhone = editPersonDescriptor.getPhone().orElseGet(personToEdit::getPhone);
-        Email updatedEmail = editPersonDescriptor.getEmail().orElseGet(personToEdit::getEmail);
-        Address updatedAddress = editPersonDescriptor.getAddress().orElseGet(personToEdit::getAddress);
-        Set<Tag> updatedTags = editPersonDescriptor.getTags().orElseGet(personToEdit::getTags);
+        Name name = editPersonDescriptor.getName();
+        Name updatedName = name == null ? personToEdit.getName() : name;
```
Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Remember your brackets arh `(name == null)`.

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/718863661aa409154218e0ecd857ed3e4480065a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a300eb7b7798dfd52a5b056874219042c7123ec7...718863661aa409154218e0ecd857ed3e4480065a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 2]
```diff
@@ -91,11 +90,16 @@ private static Person createEditedPerson(ReadOnlyPerson personToEdit,
                                              EditPersonDescriptor editPersonDescriptor) {
         assert personToEdit != null;
 
-        Name updatedName = editPersonDescriptor.getName().orElseGet(personToEdit::getName);
-        Phone updatedPhone = editPersonDescriptor.getPhone().orElseGet(personToEdit::getPhone);
-        Email updatedEmail = editPersonDescriptor.getEmail().orElseGet(personToEdit::getEmail);
-        Address updatedAddress = editPersonDescriptor.getAddress().orElseGet(personToEdit::getAddress);
-        Set<Tag> updatedTags = editPersonDescriptor.getTags().orElseGet(personToEdit::getTags);
+        Name name = editPersonDescriptor.getName();
+        Name updatedName = (name == null) ? personToEdit.getName() : name;
```
Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: `Name updatedName = editPersonDescriptor.hasName() ? editPersonDescriptor.getName() : presonToEdit.getName();`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 3]
```diff
@@ -143,46 +147,46 @@ public EditPersonDescriptor(EditPersonDescriptor toCopy) {
          * Returns true if at least one field is edited.
          */
         public boolean isAnyFieldEdited() {
-            return CollectionUtil.isAnyPresent(this.name, this.phone, this.email, this.address, this.tags);
+            return CollectionUtil.isAnyNonNull(this.name, this.phone, this.email, this.address, this.tags);
         }
 
-        public void setName(Optional<Name> name) {
+        public void setName(Name name) {
             this.name = requireNonNull(name);
```
Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: No longer should be `requireNonNull(name)`, because `null` is now a valid value.

File: src/main/java/seedu/address/commons/util/CollectionUtil.java [hunk 4]
```diff
@@ -14,23 +16,24 @@
...
      */
     public static void requireAllNonNull(Collection<?> items) {
-        Objects.requireNonNull(items);
+        requireNonNull(items);
         items.forEach(Objects::requireNonNull);
     }
 
     /**
-     * Returns true is any of the given items are present.
+     * Returns true is {@code items} contain any elements that are non-null.
```
Comment (limmlingg) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: Returns true **if**...

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 5]
```diff
@@ -200,11 +204,13 @@ public boolean equals(Object other) {
 
             // state check
             EditPersonDescriptor e = (EditPersonDescriptor) other;
-            return name.equals(e.getName())
-                    && phone.equals(e.getPhone())
-                    && email.equals(e.getEmail())
-                    && address.equals(e.getAddress())
-                    && tags.equals(e.getTags());
+
+            boolean isNameEquals = (name == null) ? (e.getName() == null) : name.equals(e.getName());
```
Comment (limmlingg) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: `areNamesEqual` might be a better name. Same for the other fields.

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 2 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: As discussed, perhaps `EditPersonDescriptor#getName` can return an `Optional` so that we can chain it.

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 6]
```diff
@@ -50,11 +50,11 @@ public EditCommand parse(String args) throws ParseException {
 
         EditPersonDescriptor editPersonDescriptor = new EditPersonDescriptor();
         try {
-            editPersonDescriptor.setName(ParserUtil.parseName(preambleFields.get(1)));
-            editPersonDescriptor.setPhone(ParserUtil.parsePhone(argMultimap.getValue(PREFIX_PHONE)));
-            editPersonDescriptor.setEmail(ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL)));
-            editPersonDescriptor.setAddress(ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS)));
-            editPersonDescriptor.setTags(parseTagsForEdit(argMultimap.getAllValues(PREFIX_TAG)));
+            ParserUtil.parseName(preambleFields.get(1)).ifPresent(editPersonDescriptor::setName);
```
Comment (limmlingg) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: @damithc Do you think we should add comments for these few lines of code, as it may not be extremely intuitive to students reading it for the first time?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: perhaps I should add a link to javadocs here HAHA:>

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/6eb5b41421395835c5dcc1593c3cc58dffb5feb5)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8e4f02b5ad8da2c440f262c2787f24f824978e4c...6eb5b41421395835c5dcc1593c3cc58dffb5feb5)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/0bdd68b171729fd378e1c01fe9d52c7fdf17d0d1~1...0bdd68b171729fd378e1c01fe9d52c7fdf17d0d1))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/88c67b5c4220bfaa3e3484868eb141fd00fc894a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8e4f02b5ad8da2c440f262c2787f24f824978e4c...88c67b5c4220bfaa3e3484868eb141fd00fc894a)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/9f83e521298febe44c8607c4ff5c496a6dc0492a~1...9f83e521298febe44c8607c4ff5c496a6dc0492a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent @limmlingg gentle reminder TO REVIEW STOP MAKING ME REMIND YOU :>

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java [hunk 7]
```diff
@@ -74,6 +74,13 @@ public void requireAllNonNullCollection() {
     }
 
     @Test
+    public void isAnyNonNull() {
+        assertFalse(CollectionUtil.isAnyNonNull());
+        assertTrue(CollectionUtil.isAnyNonNull(new Object()));
+        assertTrue(CollectionUtil.isAnyNonNull(new Object(), null));
```
Comment (yamgent) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: Should we do test for one `null` and two `null`s as well (boundary value + partition)? I.e.
`assertFalse(CollectionUtil.isAnyNonNull(null));`
`assertFalse(CollectionUtil.isAnyNonNull(null, null));`

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java (hunk 6 above)
Comment (yamgent) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: @damithc Is this use case of functional programming here appropriate, or should this be commented/simplified for student's sake since it may not come across as intuitive initially?

File: src/test/java/seedu/address/commons/util/CollectionUtilTest.java (hunk 7 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/commons/util/CollectionUtilTest.java]: As discussed,
`assertFalse(CollectionUtil.isAnyNonNull(null));` doesn't work as it will result in `items` == null, which will fail the `requireNonNull(items)`.

As such, I will be doing `assertFalse(CollectionUtil.isAnyNonNull((Object) null));` This results in `items` being an array of size 1, and `items[0] == null`.

**Comment by damithc** [comment]
> @damithc Is this use case of functional programming here appropriate, or should this be commented/simplified for student's sake since it may not come across as intuitive initially?

It is a bit obscure, but  I guess we can leave it as it is. If at all, we can add a comment at the start of the code block.

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/d0c14e3d029c79c55ba4ff1a9aa3184dd9a265a9)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a76a2a505499c1c7c68a22083026f286f4f85592...d0c14e3d029c79c55ba4ff1a9aa3184dd9a265a9)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/979c97c6eeb40c5ba4af52376dd3beb7bccb3747~1...979c97c6eeb40c5ba4af52376dd3beb7bccb3747))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent @limmlingg gentle reminder FOR YOUR REVIEW RITE NOW

File: src/main/java/seedu/address/commons/util/CollectionUtil.java [hunk 8]
```diff
@@ -14,23 +16,24 @@
...
 
     /**
-     * Returns true is any of the given items are present.
+     * Returns true if {@code items} contain any elements that are non-null.
      */
-    public static boolean isAnyPresent(Optional<?>... items) {
-        return Stream.of(items).anyMatch(Optional::isPresent);
+    public static boolean isAnyNonNull(Object... items) {
+        requireNonNull(items);
+        return Arrays.stream(items).filter(Objects::nonNull).count() > 0;
```
Comment (damithc) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: I wonder if you can use `Objects::requireNonNull` inside the lambda too (instead of counting the number of null objects).

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 9]
```diff
@@ -203,11 +203,14 @@ public boolean equals(Object other) {
...
-                    && address.equals(e.getAddress())
-                    && tags.equals(e.getTags());
+
+            boolean areNamesEqual = !getName().isPresent() ? !e.getName().isPresent() : name.equals(e.name);
+            boolean arePhonesEqual = !getPhone().isPresent() ? !e.getPhone().isPresent() : phone.equals(e.phone);
+            boolean areEmailsEqual = !getEmail().isPresent() ? !e.getEmail().isPresent() : email.equals(e.email);
+            boolean areAddressesEqual = !getAddress().isPresent()
+                    ? !e.getAddress().isPresent() : address.equals(e.address);
+            boolean areTagsEqual = !getTags().isPresent() ? !e.getTags().isPresent() : tags.equals(e.tags);
+            return areNamesEqual && arePhonesEqual && areEmailsEqual && areAddressesEqual && areTagsEqual;
```
Comment (damithc) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: This is not pretty :-p Can we write something like this?
```java
return isEqual(this, e, Person::getName) 
        && isEqual(this, e, Person::getPhone) 
        ...
```
Or something of the sort? I wouldn't even mind the following because Optional parameters are probably justified here.
```java
return isEqual(getName(), e.getName) 
        && isEqual(getPhone(), e.getPhone) 
        ...
```

File: src/main/java/seedu/address/commons/util/CollectionUtil.java (hunk 8 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: We can't, as `Objects::requireNonNull` will throw NPE instead. Unless we catch NPE, but we shouldn't do that right? :>

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/2d1151682a776c9d8a52434fdac31fef81ae77c0)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/1bea354f06cfe721d17bdf267a560a7131aa3cae...2d1151682a776c9d8a52434fdac31fef81ae77c0)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/6df7d5050b12e8fb8d826690ea6ca720dd1f8a33~1...6df7d5050b12e8fb8d826690ea6ca720dd1f8a33))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 9 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: I guess I wasn't thinking straight when I coded this ><

File: src/main/java/seedu/address/commons/util/CollectionUtil.java [hunk 10]
```diff
@@ -14,23 +16,24 @@
...
     }
 
     /**
-     * Returns true is any of the given items are present.
+     * Returns true if {@code items} contain any elements that are non-null.
      */
-    public static boolean isAnyPresent(Optional<?>... items) {
-        return Stream.of(items).anyMatch(Optional::isPresent);
+    public static boolean isAnyNonNull(Object... items) {
+        requireNonNull(items);
```
Comment (chao1995) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: Is this check redundant? Since even in the case where no arguments are given, `items` will still be an empty array rather than null?

File: src/main/java/seedu/address/commons/util/CollectionUtil.java (hunk 8 above)
Comment (chao1995) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: @Zhiyuan-Amos I wonder if we can do something like `Stream.of(items).anyMatch(Objects::nonNull);`.

File: src/main/java/seedu/address/commons/util/CollectionUtil.java (hunk 10 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: There's a possibility that `items` can be `null.` You are right in that if no arguments are given, `items` will be an empty array. However, if you pass in `(Object[]) null`, `items` will be `null`. :)

File: src/main/java/seedu/address/commons/util/CollectionUtil.java (hunk 8 above)
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: @chao1995 yup your suggestion is cleaner, thanks! :)

**Comment by CanIHasReview** [comment]
# v6

@Zhiyuan-Amos submitted v6 for review.

* [v6 1/1] [Change EditPersonDescriptor’s parameters to be non\-Optional](https://github.com/se-edu/addressbook-level4/pull/482/commits/ab9c43f623e689481d92a64b7e8fd8d1ea2a2914)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6ec9b6811339fd6a8d8a6e951b283d5e28fe6f6b...ab9c43f623e689481d92a64b7e8fd8d1ea2a2914)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/f84937557d7c4cedd79cffe2ad77a0ab75ed5242~1...f84937557d7c4cedd79cffe2ad77a0ab75ed5242))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/482/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/commons/util/CollectionUtil.java (hunk 10 above)
Comment (chao1995) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: > if you pass in (Object[]) null, items will be null. :)

Technically it's correct. But when someone does pass in a null, they are expecting the method to return false, rather than throwing an NPE?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: Hmm the rationale for putting that is:
1. The above helper method `#requireAllNonNull(Object...)` and `#requireAllNonNull(Collection<?>)` also check that the parameter passed in is non-null.
2. varargs work similarly to an array, so for example if the function is `isAnyNonNull(Object[] items)`, it makes sense to ensure that `items` itself isn't null. With the similar reasoning, if you pass in `(Object[]) null`, it seems reasonable to me to ensure that `items` itself isn't null :)

@yamgent which one would you prefer?

Comment (chao1995) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: There is the principle of least astonishment. When users call `requireNonNull`, they are prepared to get an exception when the requirement is not satisfied. But when users call `isNonNull`, they are merely asking a question, expecting the response to be either yes or no.

However, I agree that in most cases it's good to ensure the arguments passed in are not `null`, but is it possible that we also return false for this case, rather than throwing an exception?

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/commons/util/CollectionUtil.java]: Your explanation sounds good, thanks for quoting the Principle of Least Astonishment, learnt something new today :>

**Comment by Zhiyuan-Amos** [comment]
@chao1995 i've made the necessary changes already. @damithc for your review too prof.

**Comment by Zhiyuan-Amos** [comment]
@damithc gentle reminder for your review :)
